kotlinlib/src/mill/kotlinlib/android/AndroidHiltSupport.scala

```diff
@@ -0,0 +1,305 @@
+    val classes = Task.dest / "classes"
+    os.makeDir.all(classes)
+
+    val processedRoots = os.walk(
+      hiltGeneratedClasses.daggerSources.path / "hilt/internal/processedrootsentinel"
+    ).filter(_.ext == "java")
+
+    val componentTreeDeps = os.walk(hiltGeneratedClasses.javaSources.path)
+      .filter(_.toString.endsWith("_ComponentTreeDeps.java"))
+
+    val sourcesToCompile = processedRoots ++ componentTreeDeps
+
+    T.log.info(s"Compiling ${sourcesToCompile.mkString(" ")} java sources to ${classes}")
+
+    val result = worker.compileJava(
+      upstreamCompileOutput = upstreamCompileOutput(),
+      sources = sourcesToCompile,
+      compileClasspath = compileCp,
+      javacOptions = hiltJavacOptions,
+      reporter = T.ctx().reporter(hashCode()),
+      reportCachedProblems = zincReportCachedProblems(),
+      incrementalCompilation = true
+    )
+
+    val mergedClasses = Task.dest / "merged_classes"
+    os.makeDir(mergedClasses)
+
+    def merge(path: os.Path, mergedClasses: os.Path = mergedClasses): Unit =
+      os.copy(path, mergedClasses, mergeFolders = true, replaceExisting = true)
+
+    val tranformedAsmClasses = androidHiltTransformAsm()
+
+    merge(tranformedAsmClasses.kotlinCompiled.path)
+
+    merge(
+      hiltGeneratedClasses.hiltAggregatedDepsCompiled.path,
+      mergedClasses / "hilt_aggregated_deps"
+    )
+
+    merge(hiltGeneratedClasses.daggerSources.path, mergedClasses / "dagger")
+
+    merge(result.get.classes.path)
+
+    merge(hiltGeneratedClasses.javaCompiled.path)
+
+    Seq(PathRef(mergedClasses))
+
+  }
+
+  def androidHiltModule = ModuleRef(AndroidHiltModule)
+
+  /**
+   * Transforms the Kotlin classes with Hilt dependency injection context
+   *  and returns the new path of the kotlin compiled classpath. This uses
+   *  the [mill.main.android.hilt.AndroidHiltTransformAsm] that uses
+   *  the hilt gradle plugin and the android build tools
+   */
+  def androidHiltTransformAsm: T[HiltGeneratedClasses] = Task {
+    val hiltGeneratedSources = androidJavaCompileKSPGeneratedSources()
+    val kotlinCompiledClassesDir = hiltGeneratedSources.kotlinCompiled.path
+
+    val transformedClasses = Task.dest / "transformed_asm_classes"
+
+    os.makeDir.all(transformedClasses)
+    val mainClass = "mill.main.android.hilt.AndroidHiltTransformAsm"
+
+    val classPath = androidHiltModule().toolsClasspath().map(_.path)
+
+    mill.util.Jvm.callProcess(
+      mainClass = mainClass,
+      classPath = classPath,
+      mainArgs = Seq(kotlinCompiledClassesDir.toString, transformedClasses.toString)
+    )
+
+    hiltGeneratedSources.copy(
+      kotlinCompiled = PathRef(transformedClasses)
+    )
+
+  }
+
+  override def androidGeneratedCompiledClasses: T[Seq[PathRef]] =
+    androidHiltGeneratedClasses
+
+  def hiltProcessorClasspath: T[Seq[PathRef]] = Task {
+    kspApClasspath() ++ compileClasspath()
+  }
+
+  override def kotlinPrecompiledClasses: Task[Seq[PathRef]] = androidGeneratedCompiledClasses
```

Comment:
Don't reuse tasks, this results in invalid (misleading) task contexts. Instead, refer to their results.

```suggestion
  override def kotlinPrecompiledClasses: Task[Seq[PathRef]] = androidGeneratedCompiledClasses()
```

of 


```suggestion
  override def kotlinPrecompiledClasses: Task[Seq[PathRef]] = Task {
    androidGeneratedCompiledClasses()
  }
```